### PR TITLE
Refactoring VAE model

### PR DESCRIPTION
**Pull request by matsueushi**
This pull request is intend to update VAE model to follow the modern style used by lenet_mnist and dcgan_mnist. 

* Progress bars
* TensorBoard logging
* Saving model as a BSON
* Visualization of original/reconstructed image

Original
original

5 epochs
epoch_5

10 epochs
epoch_10

20 epochs
epoch_20

I also added a recipe (vae_plot.jl) for
* Loading model
* Scatter plot of training samples
* Visualization of the latent manifold.

clustering

manifold

### CODE REVIEW

File: vision/vae_mnist/vae_mnist.jl [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+using Parameters: @with_kw
+using ProgressMeter: Progress, next!
+using TensorBoardLogger: TBLogger, tb_overwrite
+using Random
+
+# load MNIST images and return loader
+function get_data(batch_size)
+    xtrain, ytrain = MLDatasets.MNIST.traindata(Float32)
+    # MLDatasets uses HWCN format, Flux works with WHCN 
+    xtrain = reshape(permutedims(xtrain, (2, 1, 3)), 28^2, :)
```
Comment (CarloLucibello) [inline vision/vae_mnist/vae_mnist.jl:26]: it turns out that MLDatasets already returns in WHCN so no need for `permutedims` here.
Images instead works with HWCN, so the permutedims should go into `convert_to_image`. 
Could you fix this also for the DCGAN and LeNet scripts?

File: vision/vae_mnist/vae_mnist.jl [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    μ, logσ, decoder(z)
+end
+
+function model_loss(encoder, decoder, λ, x, device)
+    μ, logσ, decoder_z = reconstuct(encoder, decoder, x, device)
+    len = size(x)[end]
+    # KL-divergence
+    kl_q_p = 0.5f0 * sum(@. (exp(2f0 * logσ) + μ^2 -1f0 - 2f0 * logσ)) / len
+
+    logp_x_z = -sum(binarycrossentropy.(decoder_z, x)) / len
```
Comment (CarloLucibello) [inline vision/vae_mnist/vae_mnist.jl:63]: logitbinarycrossentropy here?

Comment (CarloLucibello) [reply vision/vae_mnist/vae_mnist.jl:63]: sorry, never mind, you have a simgoid as the last activation. If you experience some  nnumerical instabilities, you could try with removing the sigmoid and using logitbinarycorssentropy instead

File: vision/vae_mnist/vae_mnist.jl (hunk 1 above)
Comment (matsueushi) [reply vision/vae_mnist/vae_mnist.jl]: Thanks, I removed `permutedims` from `get_data`.

File: vision/vae_mnist/vae_mnist.jl (hunk 2 above)
Comment (matsueushi) [reply vision/vae_mnist/vae_mnist.jl]: I switched sigmoid + bce to logitbce and inserted sigmoid to `convert_to_image`.

**Comment by CarloLucibello** [comment]
thank you, this is quite useful

**Comment by MikeInnes** [comment]
I'm not sure that all these refactorings are an improvement. The VAE example use to be very concise and clearly show the mathematical intent. Now that's obscured by a bunch of extra features. That might be worthwhile if this was code in a package, but it seems strange for a pedagogical example to spend so much time on e.g. argument handling.

**Comment by dhairyagandhi96** [comment]
One model (a new basic model) that shows off these things is fine. Doing it to everything makes it difficult to approach from a new comers standpoint.

**Comment by MikeInnes** [comment]
Yeah – model loading/saving might make sense as a tutorial for example. But all the keyword argument handling makes very little sense to me – these things are meant to be self contained.
